packages/jaeger-ui/src/model/ddg/types.tsx

```diff
@@ -53,3 +55,44 @@ export type TDdgModel = {
 };
 
 export type TDdgVertex = TVertex<{ service: string; operation: string }>;
+
+export type TDdgStateEntry =
+  | {
+      state: typeof fetchedState.LOADING;
+    }
+  | {
+      error: ApiError;
+      state: typeof fetchedState.ERROR;
+    }
+  | {
+      model: TDdgModel;
+      state: typeof fetchedState.DONE;
+      styleStates: Map<number, number>;
+    };
+
+export const stateKey = (service: string, operation: string = '*', start: number, end: number): string =>
+  [service, operation, start, end].join('\t');
+
+export type TDdgState = Record<string, TDdgStateEntry>;
+
+export enum EViewModifier {
+  None,
+  Hovered,
+  Selected,
+  Emphasized = 1 << 2, // eslint-disable-line no-bitwise
+}
+
+export type TDdgRequiredUrlState = {
+  service: string;
+  operation?: string;
+  start: number;
+  end: number;
+};
+
+export type TDdgActionMeta = {
+  query: TDdgRequiredUrlState;
+};
+
+export type TDdgAddStylePayload = TDdgRequiredUrlState & { visibilityIndices: number[]; style: number };
```

Comment:
The reason to prefer `visIndices` is that if we want to modify the view of edges (such as all edges on all paths that go through a specific vertex), it will be easiest to do so through knowing with `pathElem`s have had view modifications. Even if both the `to` and the `from` of an edge are styled, that doesn't mean the edge should be styled.

Reply:
I agree that edge `{from: a, to: b}` should not necessarily be highlighted even if both `a` and `b` are highlighted.

OTOH, if I mouseover a node, the node should trigger an action to have a view modifier added to it. But, the node only knows the [`vertexKey`](https://github.com/everett980/jaeger-ui/pull/6/files#diff-864543d0f6b69fef3144e229105f0bcdR32):

```ts
type TProps = {
  vertexKey: string;
  service: string;
  operation: string | null;
  focalNodeUrl: string | null;
  isFocalNode: boolean;
  viewModifiers: number;
  setViewModifier: (vertexKey: string, viewModifier: EViewModifier, isEnabled: boolean) => void;
};
```

So, either the reducer or some parent component needs to translate from `vertexKey` to `visIndices`. I would argue it's fine for this to happen in the reducer.

This brings up another point, though, we currently do not have a way to identify edges. `visIndices` suffers from your same point: edge `{from: a, to: b}` should not necessarily be highlighted even if both `a` and `b` are highlighted.